design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,218 @@
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and each
+setting includes the following data:
+
+- Name
+- Value
+- Registry information
+  - Description
+  - Type (e.g. enum, string, integer, or boolean)
+  - ReadOnly (a boolean indicating whether the attribute is read only)
+  - Immutable (a boolean indicating whether or not the attribute is changeable)
+  - Help text (localized)
+  - Allowable values for enum type attributes
+  - Minimum and maximum values (for numeric attributes)
+  - Minimum and maximum character lengths (for string attributes)
+
+Upon retrieving the BIOS Settigs from Ironic, the baremetal-operator will
+populate the 'currentBIOSSettings` field for each host, represented as a
+ConfigMap with an annotation in BareMetalHostStatus. As there are no fields
+in BIOS configuration that are secret (i.e. no passwords etc.) this new entry
+does not need to be secret-protected. 
+
+From `currentBIOSSettings` the BMO will also generate the
+`updateBIOSSettingsTemplate` which is a list consisting only of `Name`/`Value`
+pairs per Setting. The BMO will use the `ReadOnly` field in the Registry
+information to build this list. Settings with `ReadOnly` set to `True` will be
+ignored to avoid adding these settings via Ironic as they can't be written to
+a BMC.  'updateBIOSettingsTemplate' will be stored in the CRD, referenced in
+the BareMetalHostStatus.
+
+A user can create 'updateBIOSSettings' by copying the contents of
+'updateBIOSSettingsTemplate' to it and updating any values that need changing.
+Likewise, 'updateBIOSSettingsTemplate' can be copied to 'updateBIOSSettings'
+on multiple hosts of the same type to effectively clone the settings from a
+reference host, ensuring that all similar systems have the same BIOS
+configuration.
+
+The baremetal-operator will detect when 'updateBIOSSettings' changes by
+comparing the 'Name'/'Value' pairs in 'updateBIOSSettings' to the current BIOS
+configuration in 'currentBIOSSettings'. When a change is detected, the BMO will
+add the new values to the Ironic clean-steps API. Before adding it though, the
+BMO will do validation checks on the new values using the Registry information
+in 'currentBIOSSettings'. For example, integer types will checked against
+minimum and maximum parameters and enumeration types will be compared against
+allowable values. If the validation fails a message will be logged and the
+BIOS Setting will not be added to clean-steps.
+
+When settings are added to the Ironic clean-steps API Ironic, Ironic will set
+the BIOS Configuration in the BMC as part of the manual cleaning process which
+requires a reboot. Manual clean occurs in the Preparing state of the
+BaremetalHost, the Host will re-enter this state from Ready/Available state
+whenever its config differs from the last one it stored.
```

Comment:
I guess this means we'll have a second copy of the settings in the status (this one is a copy of the settings that we applied, whereas the other one contains all the settings we read from ironic, and they're not guaranteed to be the same).
It might be theoretically possible to collapse this into one, but not without significant and risky changes to the way we trigger manual cleaning now. We'll have to be careful with naming to minimise the confusion.

Reply:
The way I revised it is to have one copy of the data from Ironic and one another of the settings to apply. If reduces complexity to add a 2nd copy of the settings in the status to save in case of write failures we can do that, but yeah the naming will be important.